File: 01_CODE/app/src/main/java/com/beigel/leetSpeak_Generator/LeetSelectorAdapter.java
package com.beigel.leetSpeak_Generator;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.card.MaterialCardView;

import java.util.List;

public class LeetSelectorAdapter extends RecyclerView.Adapter<LeetSelectorAdapter.EnhancedLeetViewHolder> {

    private List<LeetOption> leetOptions;
    private final OnLeetSelectedListener listener;

    public interface OnLeetSelectedListener {
        void onLeetSelected(LeetOption leetOption);
        void onLeetPreview(LeetOption leetOption);
        void onEditLeet(LeetOption leetOption);
        void onToggleFavorite(LeetOption leetOption);
        void onQuickTest(LeetOption leetOption);
        void onShowTable(LeetOption leetOption); // Neue Methode für Tabellen-Anzeige
    }

    public LeetSelectorAdapter(List<LeetOption> leetOptions, OnLeetSelectedListener listener) {
        this.leetOptions = leetOptions;
        this.listener = listener;
    }

    public void updateOptions(List<LeetOption> newOptions) {
        this.leetOptions = newOptions;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public EnhancedLeetViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_leet_selector, parent, false);
        return new EnhancedLeetViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull EnhancedLeetViewHolder holder, int position) {
        LeetOption option = leetOptions.get(position);
        holder.bind(option, listener);
    }

    @Override
    public int getItemCount() {
        return leetOptions.size();
    }

    static class EnhancedLeetViewHolder extends RecyclerView.ViewHolder {
        private final MaterialCardView cardView;
        private final ImageView iconLeet;
        private final TextView textLeetName;
        private final TextView textLeetDescription;
        private final TextView textPreview;
        private final ImageView iconSelected;
        private final ImageView iconFavorite;
        private final MaterialButton buttonEdit;
        private final MaterialButton buttonQuickTest;
        private final MaterialButton buttonShowTable; // Neuer Button für Tabelle
        private final View previewSection;

        public EnhancedLeetViewHolder(@NonNull View itemView) {
            super(itemView);
            cardView = (MaterialCardView) itemView;
            iconLeet = itemView.findViewById(R.id.iconLeet);
            textLeetName = itemView.findViewById(R.id.textLeetName);
            textLeetDescription = itemView.findViewById(R.id.textLeetDescription);
            textPreview = itemView.findViewById(R.id.textPreview);
            iconSelected = itemView.findViewById(R.id.iconSelected);
            iconFavorite = itemView.findViewById(R.id.iconFavorite);
            buttonEdit = itemView.findViewById(R.id.buttonEdit);
            buttonQuickTest = itemView.findViewById(R.id.buttonQuickTest);
            buttonShowTable = itemView.findViewById(R.id.buttonShowTable); // Neue Referenz
            previewSection = itemView.findViewById(R.id.previewSection);
        }

        public void bind(LeetOption option, OnLeetSelectedListener listener) {
            // Basis-Informationen setzen
            iconLeet.setImageResource(option.getIconResId());
            textLeetName.setText(option.getName());
            textLeetDescription.setText(option.getDescription());

            // Preview aktualisieren
            String preview = generatePreview(option);
            textPreview.setText(preview);

            // Selected State
            iconSelected.setVisibility(option.isSelected() ? View.VISIBLE : View.GONE);

            // Favorite State
            iconFavorite.setVisibility(option.isFavorite() ? View.VISIBLE : View.GONE);

            // Edit Button nur für Custom Leets
            buttonEdit.setVisibility(option.isCustom() ? View.VISIBLE : View.GONE);

            // Card Stroke für ausgewähltes Element
            if (option.isSelected()) {
                cardView.setStrokeWidth(3);
                cardView.setStrokeColor(itemView.getContext().getResources()
                        .getColor(R.color.accent, itemView.getContext().getTheme()));
                cardView.setCardElevation(8f);
            } else {
                cardView.setStrokeWidth(1);
                cardView.setStrokeColor(android.graphics.Color.TRANSPARENT);
                cardView.setCardElevation(2f);
            }

            // Click Listeners
            cardView.setOnClickListener(v -> {
                if (listener != null) {
                    listener.onLeetSelected(option);
                }
            });

            // Hover effect for preview
            cardView.setOnFocusChangeListener((v, hasFocus) -> {
                if (hasFocus && listener != null) {
                    listener.onLeetPreview(option);
                }
            });

            buttonEdit.setOnClickListener(v -> {
                if (listener != null) {
                    listener.onEditLeet(option);
                }
            });

            buttonQuickTest.setOnClickListener(v -> {
                if (listener != null) {
                    listener.onQuickTest(option);
                    AnimationHelper.pulse(buttonQuickTest, 1);
                }
            });

            // Neuer Tabellen-Button Click Listener
            buttonShowTable.setOnClickListener(v -> {
                if (listener != null) {
                    listener.onShowTable(option);
                    AnimationHelper.pulse(buttonShowTable, 1);
                }
            });

            // Long Click für Favoriten
            cardView.setOnLongClickListener(v -> {
                if (listener != null) {
                    listener.onToggleFavorite(option);
                    AnimationHelper.pulse(iconFavorite, 1);
                }
                return true;
            });

            // Favorite Icon Click
            iconFavorite.setOnClickListener(v -> {
                if (listener != null) {
                    listener.onToggleFavorite(option);
                    AnimationHelper.pulse(iconFavorite, 1);
                }
            });
        }

        private String generatePreview(LeetOption option) {
            // Generate preview based on leet option
            switch (option.getMode()) {
                case ProfileManager.MODE_SIMPLE:
                    return LeetTranslator.translate("Hello", LeetTranslator.TranslationMode.SIMPLE, null);
                case ProfileManager.MODE_EXTENDED:
                    return LeetTranslator.translate("Hello", LeetTranslator.TranslationMode.EXTENDED, null);
                case ProfileManager.MODE_CUSTOM:
                    // This would need access to the profile repository
                    // For now, return a placeholder
                    return "H3ll0"; // Placeholder
                default:
                    return "H3110";
            }
        }
    }
}

File: 01_CODE/app/src/main/java/com/beigel/leetSpeak_Generator/LeetSelectorAdapter.kt
package com.beigel.leetSpeak_Generator

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.button.MaterialButton
import com.google.android.material.card.MaterialCardView

/**
 * Modern RecyclerView Adapter for leet selector with DiffUtil for efficient updates
 * Migrated to Kotlin with improved performance and type safety
 */
class LeetSelectorAdapter(
    private val listener: OnLeetSelectedListener
) : ListAdapter<LeetOption, LeetSelectorAdapter.LeetViewHolder>(LeetOptionDiffCallback()) {

    /**
     * Interface for handling leet selection events
     */
    interface OnLeetSelectedListener {
        fun onLeetSelected(leetOption: LeetOption)
        fun onLeetPreview(leetOption: LeetOption)
        fun onEditLeet(leetOption: LeetOption)
        fun onToggleFavorite(leetOption: LeetOption)
        fun onQuickTest(leetOption: LeetOption)
        fun onShowTable(leetOption: LeetOption)
    }

    /**
     * Constructor for backward compatibility with list parameter
     */
    constructor(
        leetOptions: List<LeetOption>,
        listener: OnLeetSelectedListener
    ) : this(listener) {
        submitList(leetOptions)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): LeetViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_leet_selector, parent, false)
        return LeetViewHolder(view, listener)
    }

    override fun onBindViewHolder(holder: LeetViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    /**
     * Updates the adapter with new options using DiffUtil
     */
    fun updateOptions(newOptions: List<LeetOption>) {
        submitList(newOptions)
    }

    /**
     * ViewHolder for leet option items
     */
    class LeetViewHolder(
        itemView: View,
        private val listener: OnLeetSelectedListener
    ) : RecyclerView.ViewHolder(itemView) {

        private val cardView: MaterialCardView = itemView as MaterialCardView
        private val iconLeet: ImageView = itemView.findViewById(R.id.iconLeet)
        internal val textLeetName: TextView = itemView.findViewById(R.id.textLeetName) // ✅ internal
        private val textLeetDescription: TextView = itemView.findViewById(R.id.textLeetDescription)
        private val textPreview: TextView = itemView.findViewById(R.id.textPreview)
        internal val iconSelected: ImageView = itemView.findViewById(R.id.iconSelected) // ✅ internal statt private
        internal val iconFavorite: ImageView = itemView.findViewById(R.id.iconFavorite) // ✅ internal statt private
        private val buttonEdit: MaterialButton = itemView.findViewById(R.id.buttonEdit)
        private val buttonQuickTest: MaterialButton = itemView.findViewById(R.id.buttonQuickTest)
        private val buttonShowTable: MaterialButton = itemView.findViewById(R.id.buttonShowTable)
        private val previewSection: View = itemView.findViewById(R.id.previewSection)

        fun bind(option: LeetOption) {
            // Set basic information
            iconLeet.setImageResource(option.iconResId)
            textLeetName.text = option.name
            textLeetDescription.text = option.description

            // Generate and set preview
            val preview = generatePreview(option)
            textPreview.text = preview

            // Update selection state
            iconSelected.visibility = if (option.isSelected) View.VISIBLE else View.GONE

            // Update favorite state
            iconFavorite.visibility = if (option.isFavorite) View.VISIBLE else View.GONE

            // Show edit button only for custom leets
            buttonEdit.visibility = if (option.isCustom) View.VISIBLE else View.GONE

            // Update card appearance based on selection
            updateCardAppearance(option.isSelected)

            // Set up click listeners
            setupClickListeners(option)
        }

        /**
         * Updates card appearance based on selection state
         * ✅ Public method für externe Aufrufe
         */
        fun updateCardAppearance(isSelected: Boolean) {
            val context = itemView.context

            if (isSelected) {
                cardView.strokeWidth = 3
                cardView.strokeColor = context.getColor(R.color.accent)
                cardView.cardElevation = 8f
            } else {
                cardView.strokeWidth = 1
                cardView.strokeColor = android.graphics.Color.TRANSPARENT
                cardView.cardElevation = 2f
            }
        }

        /**
         * Sets up all click listeners for the item
         */
        private fun setupClickListeners(option: LeetOption) {
            // Main card click
            cardView.setOnClickListener {
                listener.onLeetSelected(option)
            }

            // Long click for favorites
            cardView.setOnLongClickListener {
                listener.onToggleFavorite(option)
                AnimationHelper.pulse(iconFavorite, 1)
                true
            }

            // Favorite icon click
            iconFavorite.setOnClickListener {
                listener.onToggleFavorite(option)
                AnimationHelper.pulse(iconFavorite, 1)
            }

            // Edit button click
            buttonEdit.setOnClickListener {
                listener.onEditLeet(option)
            }

            // Quick test button click
            buttonQuickTest.setOnClickListener {
                listener.onQuickTest(option)
                AnimationHelper.pulse(buttonQuickTest, 1)
            }

            // Show table button click
            buttonShowTable.setOnClickListener {
                listener.onShowTable(option)
                AnimationHelper.pulse(buttonShowTable, 1)
            }

            // Preview section hover effect
            previewSection.setOnFocusChangeListener { _, hasFocus ->
                if (hasFocus) {
                    listener.onLeetPreview(option)
                }
            }
        }

        /**
         * Generates preview text for the leet option
         */
        private fun generatePreview(option: LeetOption): String {
            return when (option.mode) {
                LeetManager.MODE_SIMPLE -> {
                    LeetTranslator.translate("Hello", LeetTranslator.TranslationMode.SIMPLE, null)
                }
                LeetManager.MODE_EXTENDED -> {
                    LeetTranslator.translate("Hello", LeetTranslator.TranslationMode.EXTENDED, null)
                }
                LeetManager.MODE_CUSTOM -> {
                    // For custom profiles, we would need access to the profile
                    // This is a simplified version that assumes basic translation
                    "H3ll0" // Placeholder - would need actual profile data
                }
                else -> "H3110"
            }
        }
    }

    /**
     * DiffUtil callback for efficient list updates
     */
    private class LeetOptionDiffCallback : DiffUtil.ItemCallback<LeetOption>() {

        override fun areItemsTheSame(oldItem: LeetOption, newItem: LeetOption): Boolean {
            return if (oldItem.isCustom && newItem.isCustom) {
                oldItem.customIndex == newItem.customIndex
            } else {
                oldItem.mode == newItem.mode && oldItem.isCustom == newItem.isCustom
            }
        }

        override fun areContentsTheSame(oldItem: LeetOption, newItem: LeetOption): Boolean {
            return oldItem == newItem
        }

        /**
         * Optional: Define change payloads for more granular updates
         */
        override fun getChangePayload(oldItem: LeetOption, newItem: LeetOption): Any? {
            val changes = mutableListOf<String>()

            if (oldItem.isSelected != newItem.isSelected) {
                changes.add("selection")
            }

            if (oldItem.isFavorite != newItem.isFavorite) {
                changes.add("favorite")
            }

            if (oldItem.name != newItem.name) {
                changes.add("name")
            }

            return if (changes.isNotEmpty()) changes else null
        }
    }

    /**
     * Partial binding for efficient updates when using payloads
     */
    override fun onBindViewHolder(holder: LeetViewHolder, position: Int, payloads: MutableList<Any>) {
        if (payloads.isEmpty()) {
            super.onBindViewHolder(holder, position, payloads)
        } else {
            val item = getItem(position)

            @Suppress("UNCHECKED_CAST")
            val changes = payloads.flatMap { it as List<String> }

            if ("selection" in changes) {
                holder.iconSelected.visibility = if (item.isSelected) View.VISIBLE else View.GONE
                holder.updateCardAppearance(item.isSelected)
            }

            if ("favorite" in changes) {
                holder.iconFavorite.visibility = if (item.isFavorite) View.VISIBLE else View.GONE
            }

            if ("name" in changes) {
                holder.textLeetName.text = item.name
            }
        }
    }
}